### PR TITLE
[Crypto] DKG: timeout messages tests

### PR DESCRIPTION
**Pull request by tarakby**
- updates the testing function of DKG to support more concurrent tests and multiple edge cases:
  - all participants `dkg.Start()` are called in parallel
  - all 1-1 messages and broadcasts are non-blocking (separate threads).

   
- add tests for timeout-ed messages (private shares, verification vectors and complaints), to check the crypto library is handling late messages correctly. The number of nodes misbehaving is randomized.

- minor updates to the crypto library to improve robustness in some edge cases.

### CODE REVIEW

File: crypto/dkg_test.go [hunk 1]
```diff
@@ -426,15 +502,16 @@ func (proc *testDKGProcessor) invalidShareSend(dest int, data []byte) {
 
 	// check the behavior
 	var recipients int // number of recipients to send invalid shares to
-	if proc.malicious == manyInvalidShares {
+	switch proc.malicious {
```
Comment (jordanschalm) [inline crypto/dkg_test.go:505]: 👍

File: crypto/dkg_test.go [hunk 2]
```diff
@@ -181,13 +189,17 @@ func dkgCommonTest(t *testing.T, dkg int, n int, threshold int, test testCase) {
 	mrand.Seed(rand)
 	t.Logf("math rand seed is %d", rand)
 	var r1, r2 int
+	var h int // h is the offset of the honest nodes to check the final protocol results are correct
```
Comment (jordanschalm) [inline crypto/dkg_test.go:192]: If `h` is 2 and there are 4 nodes (1,2,3,4), then node 3 and 4 are honest?

File: crypto/dkg_test.go [hunk 3]
```diff
@@ -181,13 +189,17 @@ func dkgCommonTest(t *testing.T, dkg int, n int, threshold int, test testCase) {
 	mrand.Seed(rand)
 	t.Logf("math rand seed is %d", rand)
 	var r1, r2 int
```
Comment (jordanschalm) [inline crypto/dkg_test.go:191]: could you add a comment here explaining what these are?

It seems like they have different meanings depending on the test case -- maybe we could create them in the specific case they are for and give a more descriptive name specific to the testcase?

**Review by jordanschalm — APPROVED** [review]

File: crypto/dkg_test.go (hunk 2 above)
Comment (tarakby) [reply crypto/dkg_test.go:192]: Yes that's correct, I'll update the comment.

File: crypto/dkg_test.go (hunk 3 above)
Comment (tarakby) [reply crypto/dkg_test.go:191]: `r1` and `r2` is the number of malicious nodes. `r1` will behave maliciously and will be disqualified. `r2` will behave maliciously at first but will correct their behaviour later and won't be disqualified. 
I'll add more comments.

**Review by arrivets — APPROVED** [review]

**Pull request merged**